### PR TITLE
Basic Duplicate Detection in submission and workflow

### PR DESCRIPTION
**Pull request by kshepherd**
## References
* Replaces #8415
* REST Contract: https://github.com/DSpace/RestContract/pull/252
* Required Angular PR: https://github.com/DSpace/dspace-angular/pull/2749

The initial work on this feature was supported by TU Berlin. This work was supported by FHNW and ZHAW.

## Description
This feature introduces basic Duplicate Detection into DSpace submission and item links, using Solr's ability to search by levenshtein distance.

A new solr field `item_signature` is indexed for in-progress and archived items, with a normalised version of the title to allow it to be treated as a single term using the `~n` operator, where `n` is a configurable edit distance.

The potential duplicates detected by this search will be displayed to submitters or reviewers as a submission step, for their information (they could then choose to ignore the report if they know this is a new different item, or cancel their submission if there is a legitimate duplicate already present).

If a potential duplicate is in workspace, only the submitter of an item can see this information. 

Workflow reviewers are given a small hint indicating how many potential duplicates are detected for a pooled or claimed task, and directing them to the 'Edit' (form) where the step will show details. If the potential duplicate is in workflow, only a reviewer or admin will be able to see this information.

If versioning is used, only the latest version is included in comparison.

For archived items, a new /duplicates item link is available to see potential duplicates.

Draft documentation available (needs login) at Kim's wiki space: https://wiki.lyrasis.org/display/~kshepherd/Basic+Duplicate+Detection+for+DSpace+8

## Instructions for Reviewers
* Make sure you are using the angular PR https://github.com/DSpace/dspace-angular/pull/2749
* Configure submission to include the new `duplicate` step
* Enable duplication in `dspace/config/modules/duplicate-detection.cfg` and configure as desired
   ```
  duplicate.enable = true
   ```
* Re-index your site (REQUIRED for feature to work): `./dspace index-discovery -b`
* Add items with the exact same (or play around with edit distance) titles, and see what shows up in the Duplicate section 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by pnbecker** [comment]
We will create a PR with the RestContract soon.

**Comment by MW3000** [comment]
I had some success testing it. But it stumbled over a title that included a colon character:

The title I tested was "Tradition as a Stepping Stone for a Microbial Defined Water Kefir Fermentation Process: Insights in Cell Growth, Bioflavoring, and Sensory Perception"

That's the error I got:

```
2024-01-22 16:04:08,913 ERROR f1d12883-7093-4a2c-8745-3bc594684dcb a43f2af5-8125-4ed0-9912-006b76549064 org.dspace.app.rest.repository.ItemDuplicatesLinkRepository @ Search service error retrieving duplicates: Error from server at http://localhost:8983/solr/search: org.apache.solr.search.SyntaxError: Cannot parse '(item_signature:traditionasasteppingstoneforamicrobialdefinedwaterkefirfermentationprocess:insightsincellgrowth,bioflavoring,andsensoryperception~1)': Encountered " ":" ": "" at line 1, column 90.
Was expecting one of:
    <AND> ...
    <OR> ...
    <NOT> ...
    "+" ...
    "-" ...
    <BAREOPER> ...
    "(" ...
    ")" ...
    "*" ...
    "^" ...
    <QUOTED> ...
    <TERM> ...
    <FUZZY_SLOP> ...
    <PREFIXTERM> ...
    <WILDTERM> ...
    <REGEXPTERM> ...
    "[" ...
    "{" ...
    <LPARAMS> ...
    "filter(" ...
    <NUMBER> ...
```

**Comment by kshepherd** [comment]
@MW3000 thanks for the feedback, I will work on normalising / escaping reserved solr characters today

**Comment by MW3000** [comment]
For me, it only works with very short titles. As soon as I test with longer titles, there are no matches.
I followed this back to the Solr queries and it is Solr that does not returns fuzzy search matches for longer strings. Notice, that there are matches for non-fuzzy search for the longer strings. You can try this via the Solr console or via REST requests to Solr like this:

```
curl "http://localhost:8983/solr/search/select?indent=true&q=item_signature:averysimpletitle"
curl "http://localhost:8983/solr/search/select?indent=true&q=item_signature:averysimpletitle~1"
curl "http://localhost:8983/solr/search/select?indent=true&q=item_signature:cantileveredtandemstatorofmoderateaerodynamicloading"
curl "http://localhost:8983/solr/search/select?indent=true&q=item_signature:cantileveredtandemstatorofmoderateaerodynamicloading~1"
```

And those are the results, here from solr.log for a more compact view:

```
2024-01-25 15:45:48.572 INFO  (qtp1390301622-78) [   x:search] o.a.s.c.S.Request [search]  webapp=/solr path=/select params={q=item_signature:averysimpletitle&indent=true} hits=2 status=0 QTime=0
2024-01-25 15:45:52.582 INFO  (qtp1390301622-55) [   x:search] o.a.s.c.S.Request [search]  webapp=/solr path=/select params={q=item_signature:averysimpletitle~1&indent=true} hits=2 status=0 QTime=0
2024-01-25 15:45:58.829 INFO  (qtp1390301622-78) [   x:search] o.a.s.c.S.Request [search]  webapp=/solr path=/select params={q=item_signature:cantileveredtandemstatorofmoderateaerodynamicloading&indent=true} hits=3 status=0 QTime=0
2024-01-25 15:46:03.647 INFO  (qtp1390301622-55) [   x:search] o.a.s.c.S.Request [search]  webapp=/solr path=/select params={q=item_signature:cantileveredtandemstatorofmoderateaerodynamicloading~1&indent=true} hits=0 status=0 QTime=0
```

I could not find any information about general limits to Solr fuzzy search regarding the length of the string.

**Comment by kshepherd** [comment]
@MW3000 is your solr config any different from the default config supplied in the PR/main branch?
Long titles seem to work OK for me, e.g.

```
{
  "responseHeader":{
    "status":0,
    "QTime":0,
    "params":{
      "q":"item_signature:thistitleisquitelong\\:italsocontainsvarioussolrreservedcharacters\\!willitwork\\?wewantittomatchduplicatetitlesinoursystem~1",
      "indent":"true",
      "fl":"item_signature",
      "q.op":"OR",
      "fq":"item_signature:*",
      "rows":"20",
      "_":"1706228488431"}},
  "response":{"numFound":2,"start":0,"numFoundExact":true,"docs":[
      {
        "item_signature":["thistitleisquitelong:italsocontainsvarioussolrreservedcharacters!willitwork?wewantittomatchduplicatetitlesinoursystem"]},
      {
        "item_signature":["thistitleisquitelong:italsocontainsvarioussolrreservedcharacters!willitwork?wewantittomatchduplicatetitlesinoursystem"]}]
  }}
```

**Comment by MW3000** [comment]
@kshepherd I had the problem also with pure Solr queries via the Solr WebUI or the Solr REST API. So it did not matter if the items were in workspace, workflow or archive. I did those tests with a new Solr index of a copy of our production database.

I have since tested with a Solr index of a new and empty database, where I just added a few publications: In this case the feature works as described. Also long titles were matched correctly ✅

I will look into it and see if I can get it work with a larger search index and can find what might be the cause of my problems. Could you test with a large search index?

**Comment by MW3000** [comment]
@kshepherd  I tested again, also with a new copy of our production database and with long titles, and the feature works fine!

I could not reproduce my initial problem with long titles. So probably it was a configuration/import problem of my own.

**Comment by kshepherd** [comment]
Note to approvers: i re-ran codecov and it passed. i'm not sure why the checks list looks so confusing with 7 skipped tests, and it still has codecov on 'expected' even though i reran it and it passed. If i need to i'll just force push again to trigger it all to re-run.

**Review by MW3000 — APPROVED** [review]
I tested extensively and found that the feature is working as intended ✅

I skimmed all new and changed files, but I am not yet familiar enough with the DSpace architecture to decide if that's the way to go. It looks reasonable to me. I learned a lot.

**Comment by MW3000** [comment]
I tested again with the new search endpoint and it works as intended.

File: dspace/config/modules/duplicate-detection.cfg [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# stored signatures to be updated.
+##
+# Should the signature be normalised for case? Default: true
+duplicate.signature.normalise.lowercase = true
+# Should the signature be normalised for whitespace? Default: true
+# (highly recommended - if this is *not* used, some other placeholder needs to be used to force the value
+# to be treated as a single term by Lucene)
+duplicate.signature.normalise.whitespace = true
+
+# Levenshtein edit distance. Default:0 (eg. Test will match Txst but not Txxt)
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace/config/modules/duplicate-detection.cfg:19]: I have done this, good idea

File: dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+     * @return DiscoverResult as a result of performing search. Null if invalid.
+     *
+     * @throws SearchServiceException if an error was encountered during the discovery search itself.
+     */
+    @Override
+    public DiscoverResult searchDuplicates(Context context, Item item) throws SearchServiceException {
+
+        // If the item is null or otherwise invalid (template, etc) then throw an appropriate error
+        if (item == null) {
+            throw new ResourceNotFoundException("No such item: " + item);
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java]: I've updated this error message

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java [hunk 3]
```diff
@@ -365,4 +375,53 @@ protected ItemRest createAndReturn(Context context, List<String> stringList)
...
+    /**
+     * Search method to find and return a paged list of duplicates, if the current user has authorization
+     *
+     * @param itemUuid the UUID of the item for which we are searching duplicates
+     * @param optionalPageable pagination options
+     * @return pagination potential duplicate results
+     */
+    @PreAuthorize("hasPermission(#itemUuid, 'ITEM', 'READ')")
+    @SearchRestMethod(name = "findDuplicates")
+    public Page<PotentialDuplicateRest> findDuplicates(@Parameter(value = "uuid", required = true) UUID itemUuid,
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:388]: this method is not appropriate for the ItemRestRepository. You will need eventually a dedicated RestRepository to expose the concept of "duplicates" where you can provide such search method

**Review by abollini — CHANGES_REQUESTED** [review]
I provided few inline comments by code inspection, my major concern is about the divergence of this feature from what already exists in DSpace-CRIS and was originally planned to be ported to DSpace. If this is not longer the plan on the Library of Code side 4Science can consider it as a contribution for DSpace 9, see also https://github.com/DSpace/RestContract/pull/252#pullrequestreview-1896437014

**Comment by kshepherd** [comment]
As per latest situation, I am trying to refactor this to use completely different REST methods. I am having trouble as if I follow the logic here, I'm not allowed to use searchBy on the related object, or a link repository from the related object, so it seems like a plain controller is the only good fit for these non-addressable duplicates. But this does not allow registration of SearchRestMethod through annotations and I'll have to rewrite the angular data services, too.

My work (in progress, not yet committed) also reverts solr schema changes so there is no confusion or conflict with the 4science work that is coming and may supercede/replace this feature.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java (hunk 3 above)
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:388]: I have moved this to a dedicated controller.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+     * @param uuid The item UUID to search
+     * @param page Pagination options
+     * @param assembler The paged resources assembler to construct the paged model
+     * @return Paged list of potential duplicates
+     * @throws Exception
+     */
+    @RequestMapping(method = RequestMethod.GET, value = "/search")
+    @PreAuthorize("hasPermission(#uuid, 'ITEM', 'READ')")
+    public PagedModel<PotentialDuplicateResource> searchPotentialDuplicates(
+            @RequestParam(name = "uuid", required = true) UUID uuid, Pageable page, PagedResourcesAssembler assembler) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:88]: If I'm not mistaken, this method is **only called** during submission/workflow.  That makes me suspect we might want to rename this endpoint to be
```
/api/submission/duplicates/search?uuid=
```

Currently, it appears to be named:
```
/api/duplicates/search?uuid=
```

While I'm not against that name, it implies this should be used in **other features** as it's a top-level endpoint right under `/api`...instead of grouping it with other submission features under `/api/submission` (like `/api/submission/vocabularies` and similar)

File: dspace/config/item-submission.xml [hunk 5]
```diff
@@ -267,6 +274,10 @@
             <!-- To include the 'Show Identifiers' step, uncomment the line below. Make sure
                  that identifiers.cfg and identifier-service.xml are properly configured. -->
             <!--<step id="identifiers"/>-->
+            <!-- To include the 'Duplicate Detection' step, uncomment the line below. Make sure
+                 that duplicate-detection.cfg is properly configured and items have been reindexed since
+                 enabling the feature -->
+            <!--<step id="duplicates"/>-->
```
Comment (tdonohue) [inline dspace/config/item-submission.xml:280]: Tiny thing...it might be more "useful" to move this step to *after* the metadata entry steps below.  I found in testing that I didn't notice the duplicate step at first because I had to scroll back to the top of the submission form to see where it appeared.  This is an optional suggestion though as sites can move this wherever they wish.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@kshepherd : Gave this a test/review today.  Overall, the backend code looks good...but there are some areas to cleanup.  I've also run into some bugs (namely that the duplicates don't load in the workflow >edit submission steps)...but I'll report those on the frontend PR.

**Comment by MW3000** [comment]
I tested the feature again with the new changes. For me the feature worked well.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java (hunk 4 above)
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:88]: Although right now the only display of duplicates is in submission / workflow, I know of at least one group that plans to generate their own reports of archived items with duplicate titles, via REST calls, which is I wouldn't want to make it seem like this is a submission-only feature. (identitying duplicates can be a useful clean-up exercise as well as a warning during submission). Does that make sense?
I will change the path if required, just explaining my reasoning.

File: dspace-api/src/test/data/dspaceFolder/config/local.cfg [hunk 6]
```diff
@@ -174,3 +174,6 @@ authority.controlled.dspace.object.owner = true
 # Configuration required for thorough testing of browse links
 webui.browse.link.1 = author:dc.contributor.*
 webui.browse.link.2 = subject:dc.subject.*
+
+# Enable duplicate detection for tests
+duplicate.enable = true
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/test/data/dspaceFolder/config/local.cfg:179]: NOTE: I realized this *might be necessary* if you use the `@ConditionalOnProperty` approach to enable/disable your controller.  As I think that does require you enable the controller in the global `local.cfg` for the Test Environment...since the controller decides whether to load when the application boots up.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java (hunk 4 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:88]: @kshepherd : Makes sense.  I think the only reason I bring this up, is that most of our endpoints look like this:
```
/api/[category]/[endpoint-name]/search
```
But, this one looks like:
```
/api/[endpoint-name]/search
```

So, I was trying to figure out an existing "category" that it could go into.   It seemed odd to me that this endpoint is a category all to itself.  There's also the "tools" category (which is used by a few endpoints).   The "submission" one just jumped out as related since that's where it's currently used.

This isn't the highest priority thing, to be honest...I was merely looking for consistency with other REST endpoints.

File: dspace/config/item-submission.xml (hunk 5 above)
Comment (kshepherd) [reply dspace/config/item-submission.xml:280]: Good call, I agree - especially since a new submission needs metadata before a duplicate match makes sense. Users will customise this to fit their purpose anyway. I have made this change.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DuplicateDetectionRestIT.java [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+                .build();
+        WorkspaceItem workspaceItem3 = WorkspaceItemBuilder.createWorkspaceItem(context, simpleCol)
+                .withTitle(item1Title)
+                .withTitle("Public item III")
+                .withIssueDate("2013-10-17")
+                .withAuthor("Smith, Donald Y.")
+                .withSubject("ExtraEntry 3")
+                .withSubmitter(eperson)
+                .build();
+        log.error("EPERSON FULL NAME IS " + eperson.getFullName());
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/DuplicateDetectionRestIT.java:169]: Yep! Done.

File: dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+            // Not a workspace or workflow item, but is it archived, not withdrawn, and discoverable?
+            // Is it readable by the current user?
+            if (authorizeService.authorizeActionBoolean(context, resultItem, Constants.READ)) {
+                return Optional.of(potentialDuplicate);
+            }
+        } else if (authorizeService.isAdmin(context, resultItem)) {
+            // Admins can always read, return immediately
+            return Optional.of(potentialDuplicate);
+        } else {
+            log.error("No valid permission to return this item");
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java:242]: yes i changed it to info (it is quite a common occurence and not cause for concern, just means we are appropriately not including an item) and included the item_id, thanks

File: dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            workflowItem = workflowItemService.findByItem(context, resultItem);
+            workspaceItem = workspaceItemService.findByItem(context, resultItem);
+        }
+
+        // Result item must not be null, a template item, or actually identical to the original
+        if (resultItem == null) {
+            log.error("skipping null item in duplicate search results");
+            return Optional.empty();
+        } else if (resultItem.getTemplateItemOf() != null) {
+            log.error("skipping template item in duplicate search results");
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java:179]: changed this to info and included id

File: dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        if (indexableObject instanceof IndexableItem) {
+            resultItem = ((IndexableItem) indexableObject).getIndexedObject();
+            // Attempt resolution of workflow or workspace items, tested later
+            workflowItem = workflowItemService.findByItem(context, resultItem);
+            workspaceItem = workspaceItemService.findByItem(context, resultItem);
+        }
+
+        // Result item must not be null, a template item, or actually identical to the original
+        if (resultItem == null) {
+            log.error("skipping null item in duplicate search results");
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java:176]: agreed, warn is more appropriate

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * The controller for the api/duplicates endpoint, which handles requests for finding
+ * potential duplicates
+ *
+ * @author Kim Shepherd
+ */
+@RestController
+@RequestMapping("/api/" + PotentialDuplicateRest.CATEGORY)
+public class DuplicateDetectionRestController implements InitializingBean {
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:51]: Done, thanks for the tip

File: dspace-api/src/test/data/dspaceFolder/config/local.cfg (hunk 6 above)
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/test/data/dspaceFolder/config/local.cfg:179]: As per your other comment, i tested and we do need this with the ConditionalProperty annotation used in our controller

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java (hunk 3 above)
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:388]: yes, an oversight. removed.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DuplicateDetectionRestIT.java [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+                // Valid duplicates array
+                .andExpect(jsonPath("$._embedded.potentialDuplicateResources").doesNotExist());
+    }
+
+    @Override
+    public void destroy() throws Exception {
+        if (anotherEPerson != null) {
+            try {
+                context.turnOffAuthorisationSystem();
+                ePersonService.delete(context, anotherEPerson);
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/DuplicateDetectionRestIT.java:419]: done

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java (hunk 4 above)
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:91]: i have moved it to /api/submission/duplicates/search and updated names / angular paths / docs / rest contract accordingly

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java [hunk 13]
```diff
@@ -95,6 +96,9 @@ public class ItemRestRepository extends DSpaceObjectRestRepository<Item, ItemRes
     @Autowired
     private UriListHandlerService uriListHandlerService;
 
+    @Autowired
+    DuplicateDetectionService duplicateDetectionService;
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:100]: Unless I'm missing something, I think you can remove this? It doesn't appear to be used as you removed everything else from ItemRestRepository.  (But please check to see if I'm correct)

**Review by tdonohue — APPROVED** [review]
👍 @kshepherd : Thanks for the cleanup!  This looks good & works.  While I'm approving this PR, there are two **very minor** code cleanup notes below.  I think we should at least fix the `item-submission.xml` comments.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java (hunk 13 above)
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:100]: correct, another oversight not picked up in my self-review, fixed.

File: dspace/config/item-submission.xml [hunk 14]
```diff
@@ -267,6 +274,10 @@
             <!-- To include the 'Show Identifiers' step, uncomment the line below. Make sure
                  that identifiers.cfg and identifier-service.xml are properly configured. -->
             <!--<step id="identifiers"/>-->
+            <!-- To include the 'Duplicate Detection' step, uncomment the line below. Make sure
+                 that duplicate-detection.cfg is properly configured and items have been reindexed since
+                 enabling the feature -->
+            <!---->
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace/config/item-submission.xml:280]: thanks, done

**Comment by tdonohue** [comment]
@kshepherd : Just in case you missed it, this PR has a merge conflict as well (likely caused by the merger of the Content Reports backend PR #8598).

**Comment by kshepherd** [comment]
@tdonohue yes, missed a push. Thanks for all your thorough reviewing and testing, and sorry for all the bits and pieces I've missed here - this PR has changed quite a lot from its initial creation so I've dropped the ball in a few places

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+/**
+ * The controller for the api/duplicates endpoint, which handles requests for finding
+ * potential duplicates
+ *
+ * @author Kim Shepherd
+ */
+@RestController
+@ConditionalOnProperty("duplicate.enable")
+@RequestMapping("/api/" + PotentialDuplicateRest.CATEGORY + "/" + PotentialDuplicateRest.NAME)
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:52]: why a new controller is needed? it is a search method so it should be sufficient to create a REST repository following our current architecture. I strongly recommend to get ride of this controller and just introduce the PotentialDuplicateRestRepository

**Review by abollini — DISMISSED** [review]
please check my inline comment

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java (hunk 15 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:52]: @kshepherd : Just to clarify @abollini 's comment more... He's right that this code could be simplified if refactored from `DuplicateDetectionRestController` to `PotentialDuplicateRestRepository` (or whatever you want to name it).  That should not be a large refactor.

Doing that refactor would also fix the feedback @abollini noted on [his REST Contract comment](https://github.com/DSpace/RestContract/pull/252#pullrequestreview-1909161686).

Essentially, you'd just move your existing `DuplicateDectionsRestController.searchPotentialDuplicates()` method into a `PotentialDuplicateRestRepository.searchPotentialDuplicates()` that looks more like this:
```
@PreAuthorize("hasPermission(#uuid, 'ITEM', 'READ')")
@SearchRestMethod(name="byUUID")
public PagedModel<PotentialDuplicateResource> searchPotentialDuplicates(
            @RequestParam(name = "uuid", required = true) UUID uuid, Pageable page, PagedResourcesAssembler assembler) {
```

I don't know if that code is all completely correct... but that "@SearchRestMethod" annotation would result in a subpath like `/search/byUUID?uuid=` under the `/api/submission/duplicates/` path (which would be defined by `PotentialDuplicateRestRepository`).

For many examples of how this works, search for the `@SearchRestMethod` annotation in our codebase...as that is used for almost all subpath `/search` endpoints.   It can only be used in `*RestRepository` class though. 

Apologies, I should have thought of this earlier as it actually should simplify your code a bit...using a *RestRepository also ensures that, by default, you align with other REST endpoint designs.

Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:52]: I thought rest repositories were only for addressable objects.
While we do search for an item ID, we are returning stubs which are not themselves an addressable DSO or other db object, just a stub with preview values. I was told to stop using the ItemRestRepository for this very reason in the first review. I will see if it is possible to implement a DuplicateRestRepository, and will switch all my code back to using a searchBy method if that works. I may not have time for this until after the weekend.

Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:52]: @kshepherd : I think there's been a misunderstanding?  The problem with using the ItemRestRepository was never about the fact that you _shouldn't use a RestRepository.   Instead, the issue was that the ItemRestRepository class is meant to *always return/represent `Item` resources* (or objects).   In your code, you are **not** returning an Item resource, but are returning something you've named "Duplicate" (`"type": "DUPLICATE"`).   This is a **new type of resource** which is related to an Item, but it is **not** and Item (`"type": "ITEM"`).

So, the original feedback was that you shouldn't use the `ItemRestRepository` simply because your search endpoints don't return `Item` resources.   Instead, you'd need to define your own endpoint (which you've already done) for that new resource...and then create your own RestRepository **or** Controller. 

As a sidenote: In some situations, we've found that Controllers are useful as they allow more control over creating super-custom endpointss.  However, ideally, there's a preference for RestRepository classes because they "force" your endpoints to be well-structured... that "well-structuring" makes it easier to create endpoints that immediately align with our Contract. But it can be limiting for some endpoints that don't fit into this structure.

In all honesty, I should have asked initially *why* you switched to a Controller.  But, this is simply something I wasn't paying attention too as I had several other feature PRs I was attempting to help other developers along with.

Long story short, you can and should use a RestRepository class when it is possible to do so.  There are may non-addressable DSOs / db objects which already have their own RestRepository.  Here's a few: `AuthorizationRestRepository`, `BulkAccessConditionRestRepository`, `ConfigurationRestRepository`, `DiscoveryRestRepository`, ... `VocabularyRestRepository`  (None of these are DSOs or have db objects).  Generally there is a RestRepository for almost every top-level endpoint (with a few exceptions where a Controller is used instead)

I hope that helps, and apologies that I didn't mention/notice this sooner.  This is one of the reasons I always try to get a few reviewers on every PR -- I can miss things too & additional eyes always help.

Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java:52]: @tdonohue  Sorry, yes I do understand why returning a non-Item from an item repository was bad, I also didn't full explain myself properly in my last comment. I have reverted my controller and tried to adapt these requirements to suit the way duplicates work.

**Summary update**:

I have now implemented a `DuplicateRestRepository` which only implements and registers a SearchRestMethod. It returns Method Not Implemented exceptions for both `findOne` and `findAll`.

To me this is a confusing use of REST repositories -- I thought the ability to at least find an addressable object with `/api/endpointname/{identifier}` was the whole point of them.
Only classes that extend `DSpaceRestRepository<Model, ID>` are able to register `SearchRestMethod` methods so this is what I have had to extend in order to make the required refactory.

**Detailed notes to explain my thought process:**

It is the fact that I cannot use a `@SearchRestMethod` method without extending `DSpaceRestRepository`, which needs addressable objects (am I using this term wrong? "things able to be referenced using IDs" is what I mean), and that doesn't fit the way I have implemented this very simple duplicate detection.

Look at the way the examples you mention are extending `DSpaceRestRepository` -- typed for a REST model and a class of that object's addressable ID -- yes they are not DSOs and not all in the database, but the RestModels are still first-class "things" which can be individually addressed/retrieved with a String ID:

`public class BulkAccessConditionRestRepository extends DSpaceRestRepository<BulkAccessConditionRest, String>`
`public class ConfigurationRestRepository extends DSpaceRestRepository<PropertyRest, String>`
`public class VocabularyRestRepository extends DSpaceRestRepository<VocabularyRest, String>`

There is no string, UUID or integer which can retrieve a single duplicate stub, they are lists of previews generated after comparing an item with everything else in the search index. (but they are not SearchResultsRest lists, unfortunately which again require wrapping directly addressable resources).

So the only exception in that set of examples was `DiscoveryRestRepository`, which only extends `AbstractDSpaceRestRepository`, but this won't allow a `@SearchRestMethod` method to be used -- indeed the actual endpoint mapping still happens on a controller class for this example.

I went through this analysis and thinking when I was moving my code out of `ItemRestRepository` and it is for the reasons above that I thought a controller was the most appropriate solution.
However, I will try to make things work with RestRepository patterns as instructed.

Is the compromise here to "pretend" there is a String that can address a Duplicate object somehow, and then either refuse to implement findOne and findAll, or just mush this implementation later, like treat that string as a search query and return first result for findOne? Or instead use UUID and allow people to get a "duplicate" by item ID to just see what that item *would* look like rendered as a duplicate..?

To avoid confusion I am going for the first option - I will tell DSpaceRestRepository that this extension uses a 'String' ID, but I will throw NOT_IMPLEMENTED exceptions for both `findOne` and `findAll`, and just implement a `SearchRestMethod` as instructed.

**Comment by kshepherd** [comment]
force pushed an update with latest required changes. (coar notify and other merges make this heavy with conflicts)

**Review by tdonohue — APPROVED** [review]
👍 Thanks @kshepherd !  Tested this today with the frontend PR.  It works, and the code looks good.  I've verified also that @abollini 's feedback has been addressed...the endpoint is renamed & made into a RestRepository.

Any further changes can be provided as bug-fix PRs (if necessary).  This seems ready to merge.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java (hunk 15 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java]: @kshepherd : Thanks for explaining your point better. I now understand why you were concerned... that said, I like the approach you took.  I think it's OK to have a RestRepository which only implements a `@SearchRestMethod`...and I agree that, in that scenario, the unimplemented methods should be documented to not exist (and throw errors).  Thanks!
